cpp/src/rolling/rolling_detail.cuh

```diff
@@ -64,322 +66,733 @@
-                               thrust::make_counting_iterator(end_index),
-                               [&input](auto i) { return input.is_valid_nocheck(i); });
+           // SUM supports all fixed width types except timestamps
+           ((op == aggregation::SUM) && (cudf::is_fixed_width<T>() && !cudf::is_timestamp<T>())) ||
+
+           // MEAN supports numeric and duration
+           ((op == aggregation::MEAN) && (cudf::is_numeric<T>() || cudf::is_duration<T>()));
+  }
+
+  // operations we do support
+  template <typename T = InputType, aggregation::Kind O = op>
+  DeviceRolling(size_type _min_periods, typename std::enable_if_t<is_supported<T, O>()>* = nullptr)
+    : min_periods(_min_periods)
+  {
+  }
+
+  // operations we don't support
+  template <typename T = InputType, aggregation::Kind O = op>
+  DeviceRolling(size_type _min_periods, typename std::enable_if_t<!is_supported<T, O>()>* = nullptr)
+    : min_periods(_min_periods)
+  {
+    CUDF_FAIL("Invalid aggregation/type pair");
+  }
+
+  // perform the windowing operation
+  template <typename OutputType, bool has_nulls, typename T = InputType, aggregation::Kind O = op>
+  std::enable_if_t<is_supported<T, O>(), bool> __device__
+  operator()(column_device_view const& input,
+             column_device_view const& ignored_default_outputs,
+             mutable_column_device_view& output,
+             size_type start_index,
+             size_type end_index,
+             size_type current_index)
+  {
+    using AggOp = typename corresponding_operator<O>::type;
+    AggOp agg_op;
+
+    // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
+    // for CUDA 10.0 and below (fixed in CUDA 10.1)
+    volatile cudf::size_type count = 0;
+    OutputType val                 = AggOp::template identity<OutputType>();
+
+    for (size_type j = start_index; j < end_index; j++) {
+      if (!has_nulls || input.is_valid(j)) {
+        OutputType element = input.element<device_storage_type_t<InputType>>(j);
+        val                = agg_op(element, val);
+        count++;
+      }
     }
-    output.element<OutputType>(current_index) = count;
+
+    bool output_is_valid = (count >= min_periods);
+
+    // store the output value, one per thread
+    cudf::detail::rolling_store_output_functor<OutputType, op == aggregation::MEAN>{}(
+      output.element<OutputType>(current_index), val, count);
+
+    return output_is_valid;
   }
 
-  return output_is_valid;
-}
+  template <typename OutputType, bool has_nulls, typename T = InputType, aggregation::Kind O = op>
+  std::enable_if_t<!is_supported<T, O>(), bool> __device__
+  operator()(column_device_view const& input,
+             column_device_view const& ignored_default_outputs,
+             mutable_column_device_view& output,
+             size_type start_index,
+             size_type end_index,
+             size_type current_index)
+  {
+    return false;
```

Comment:
I'm testing if the need for multiple SFINAEs (based on `is_supported<T,O>()`) can be reliably worked around with `if constexpr()`. I'll post here shortly.

Reply:
I think I have it, but it will need careful wording to work around a bug in `nvcc`:
```c++
  template <typename T = InputType>
  DeviceRollingLead(size_type _row_offset) : row_offset(_row_offset)
  {
    if constexpr (not is_supported<T>())
    {
      CUDF_FAIL("Invalid aggregation/type pair");
    }
  }

  template <typename OutputType, bool has_nulls, typename T = InputType>
  bool __device__
  operator()(column_device_view const& input,
             column_device_view const& default_outputs,
             mutable_column_device_view& output,
             size_type start_index,
             size_type end_index,
             size_type current_index)
  {
    if constexpr (is_supported<T>())
    {
      // Offsets have already been normalized.

      // Check if row is invalid.
      if (row_offset > (end_index - current_index - 1)) {
        // Invalid row marked. Use default value, if available.
        if (default_outputs.size() == 0 || default_outputs.is_null(current_index)) { return false; }

        output.element<OutputType>(current_index) =
          default_outputs.element<OutputType>(current_index);
        return true;
      }

      // Not an invalid row.
      auto index   = current_index + row_offset;
      auto is_null = input.is_null(index);
      if (!is_null) {
        output.element<OutputType>(current_index) =
          input.element<device_storage_type_t<InputType>>(index);
      }
      return !is_null;
    }

    return false;
  }
```
This is admittedly a nitpick.

Reply:
See:  the `corresponding_rolling_operator` stuff I added. It accomplishes the same thing with a tiny bit of SFINAE